File: service/api-libs/service_api_lib.py
import requests
import json
from typing import TypeVar

def raise_error(e, res):
    raise Exception(str(e) + res.json()['error'])

# Class that is used for anything that requires authorization
# Client can only be connected to one database at a time
T = TypeVar('T', bound='Client')
class Client:
    def __init__(self, name="", password="", auth_token="", url=""):
        self.name = name
        self.password = password
        self.auth_token = auth_token
        self.url = url
        
    def get(self, category, key) :
        data = {
            'name': self.name,
            'auth_token': self.auth_token,
            'category_name': category,
            'key': json.dumps(key)
        }
        return requests.post(self.url + '/db-get', json=data)
    
    def get_range(self, category, key) :
        data = {
            'name': self.name,
            'auth_token': self.auth_token,
            'category_name': category,
            'key': json.dumps(key)
        }
        return requests.post(self.url + '/db-get-range', json=data)
    
    def put(self, category, key, value) :
        data = {
            'name': self.name,
            'auth_token': self.auth_token,
            'category_name': category,
            'key': json.dumps(key),
            'value': json.dumps(value)
        }
        return requests.post(self.url + '/db-put', json=data)
    
    def create_category(self, category: str, key_schema: dict, value_schema: dict) :
        data = {
            'name': self.name,
            'auth_token': self.auth_token,
            'category_name': category,
            'key_schema': json.dumps(key_schema),
            'value_schema': json.dumps(value_schema)
        }
        return requests.post(self.url + '/db-create-category', json=data)
    
    def add_client_to_category(self, category: str, add_client: T) :
        data = {
            'name': self.name,
            'auth_token': self.auth_token,
            'category_name': category,
            'add_name': add_client.name
        }
        return requests.post(self.url + '/db-add-client', json=data)
    
    def is_authenticated(self):
        data = {
            'name': self.name,
            'auth_token': self.auth_token
        }
        return requests.post(self.url + '/db-check-authenticated', json=data)

# Class that is used for database wide operations that don't require authorization
class Buddy:
    def __init__(self, url: str, path: str = ""):
        self.path = path
        self.url = url
    
    def create(self, path: str):
        res = requests.post(self.url + '/db-create-buddy', json={"path": path})
        if res.status_code == 200 :
            self.path = res.json()['folder_path']
        return res
            
    def connect(self, path: str = ""):
        p = self.path
        if path != "":
            p = path
        return requests.post(self.url + '/db-connect-buddy', json={"path": p})

    def disconnect(self):
        return requests.get(self.url + '/db-disconnect-buddy')
            
        
    def create_client(self, client: Client):
        data = {
            'name': client.name, 
            'password': client.password
        }
        res = requests.post(self.url + '/db-create-client', json=data)
        if (res.status_code == 200):
            client.url = self.url
            client.auth_token = res.json()['auth_token']
        return res          
    
    def connect_client(self, client: Client):
        data = {
            'name': client.name, 
            'password': client.password
        }
        res = requests.post(self.url + '/db-connect-client', json=data)
        if (res.status_code == 200):
            client.url = self.url
            client.auth_token = res.json()['auth_token']
        return res
    
    def disconnect_client(self, client: Client):
        data = {
            'name': client.name, 
            'auth_token': client.auth_token
        }
        return requests.post(self.url + '/db-disconnect-client', json=data)
    
    def list_clients(self):
        return requests.get(self.url + '/db-list-clients')
                
    def list_categories(self):
        return requests.get(self.url + '/db-list-categories')
            
    def get_path(self):
        return self.path
    
    def get_url(self):
        return self.url
    
    def ping(self):
        return requests.get(self.url + '/db-ping')
    
    def is_connected(self):
        return requests.get(self.url + '/db-check-buddy-connected')
    
    
    

File: service/api_libs/service_api_lib.ts
async function handle_response(res: any): Promise<Response> {
    let response: Response = new Response(500, {});
    await res.text()
    .then(function (parsed: any) {
        let json: any;
        try {
            json = JSON.parse(parsed);
        } catch (error) {
            throw new Error("Trouble parsing following as JSON. " + parsed);
        }
        response = new Response(res.status, json);
    })
    .catch(function (err: any) {
        response = new Response(res.status, {"error": "Trouble parsing as text. " + err});
    });
    return response;
}

async function post_json(url: string, data: any): Promise<Response> {
    return await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(data),
    })
    .then(handle_response)
    .catch(function (err: any) {
        return new Response(500, {"error": "Trouble fetching. " + err});
    });
}

async function get_json(url: string): Promise<Response> {
    return await fetch(url, {
        method: 'GET'
    })
    .then(handle_response)
    .catch(function (err: any) {
        return new Response(500, err);
    });
}

export class Response {
    status_code: number;
    data: any;

    constructor(status_code: number, data: any) {
        this.status_code = status_code;
        this.data = data;
    }
}

export class Client {
    readonly name: string;
    readonly password: string;
    auth_token: string;
    url: string;

    constructor(name: string, password: string, auth_token?: string, url?: string) {
        this.name = name;
        this.password = password;
        if (typeof auth_token !== 'undefined') {
            this.auth_token = auth_token;
        } else {
            this.auth_token = '';
        }
        if (typeof url !== 'undefined') {
            this.url = url;
        } else {
            this.url = '';
        }
    }

    async is_authenticated() : Promise<Response> {
        if (this.url === '' || this.auth_token === '') {
            return new Response(200, {"data": false, "error": ""});
        }
        let data: any = {
            'name': this.name,
            'auth_token': this.auth_token
        }
        return await post_json(this.url + '/db-check-authenticated', data);
    }
    
}

export class Buddy {
    url: string;
    path: string;

    constructor(url: string, path?: string) {
        this.url = url;
        if (typeof path !== 'undefined') {
            this.path = path;
        } else {
            this.path = '';
        }
    }

    async create(path: string): Promise<Response> {
        let data: any = {
            'path': path
        }
        let res = await post_json(this.url + '/db-create-buddy', data);
        if (res.status_code === 200) {
            this.path = res.data['path'];
        }
        return res;
    }

    async connect(path?: string): Promise<Response> {
        let p = this.path;
        if (typeof path !== 'undefined') {
            p = path;
        }
        let data: any = {
            'path': p
        }
        return await post_json(this.url + '/db-connect-buddy', data);
    }

    async disconnect(): Promise<Response> {
        return await get_json(this.url + '/db-disconnect-buddy');
    }

    async create_client(client: Client): Promise<Response> {
        let data: any = {
            'name': client.name,
            'password': client.password
        }
        let res = await post_json(this.url + '/db-create-client', data);
        if (res.status_code === 200) {
            client.url = this.url;
            client.auth_token = res.data['auth_token'];
        }
        return res;
    }

    async connect_client(client: Client): Promise<Response> {
        let data: any = {
            'name': client.name,
            'password': client.password
        }
        let res = await post_json(this.url + '/db-connect-client', data);
        if (res.status_code === 200) {
            client.url = this.url;
            client.auth_token = res.data['auth_token'];
        }
        return res;
    }

    async disconnect_client(client: Client): Promise<Response> {
        let data: any = {
            'name': client.name,
            'auth_token': client.auth_token
        }
        return await post_json(this.url + '/db-disconnect-client', data);
    }

    async list_clients(): Promise<Response> {
        return await get_json(this.url + '/db-list-clients');
    }

    async list_categories(): Promise<Response> {
        return await get_json(this.url + '/db-list-categories');
    }

    get_path(): string {
        return this.path;
    }

    get_url(): string {
        return this.url;
    }

    async ping() : Promise<Response> {
        return await get_json(this.url + '/db-ping');
    }

    async is_connected(): Promise<Response> {
        return await get_json(this.url + '/db-check-buddy-connected');
    }
}